app/src/behaviors/behavior_tap_hold.c

```diff
@@ -0,0 +1,432 @@
+    LOG_ERR("unable to store tap-hold info, did you press more than %d tap-holds?", ZMK_BHV_TAP_HOLD_MAX_HELD);
+    return 0;
+  }
+
+  LOG_DBG("%d new undecided tap_hold", position);
+  undecided_tap_hold = tap_hold;
+  k_delayed_work_submit(&tap_hold->work, cfg->tapping_term_ms());
+
+  //todo: once we get timing info for keypresses, start the timer relative to the original keypress
+  // don't forget to simulate a timer-event before the event after that time was handled.
+
+  return 0;
+}
+
+static int on_tap_hold_binding_released(struct device *dev, u32_t position, u32_t _, u32_t __)
+{
+  struct active_tap_hold *tap_hold = find_tap_hold(position);
+  if(tap_hold == NULL) {
+    LOG_ERR("ACTIVE_TAP_HOLD_CLEANED_UP_TOO_EARLY");
+    return 0;
+  }
+
+  int work_cancel_result = k_delayed_work_cancel(&tap_hold->work);
+  decide_tap_hold(tap_hold, TH_KEY_UP);
+
+  struct zmk_behavior_binding *behavior;
+  if (tap_hold->is_hold) {
+    behavior = &tap_hold->config->behaviors->hold;
+  } else {
+    behavior = &tap_hold->config->behaviors->tap;
+  }
+
+  struct device *behavior_device = device_get_binding(behavior->behavior_dev);
+  behavior_keymap_binding_released(behavior_device,  tap_hold->position, behavior->param1, behavior->param2);
+
+  if(work_cancel_result == -EINPROGRESS) {
+    // let the timer handler clean up
+    // if we'd clear now, the timer may call back for an uninitialized active_tap_hold.
+    LOG_DBG("%d tap-hold timer work in event queue", position);
+    tap_hold->work_is_cancelled = true;
+  } else {
+    LOG_DBG("%d cleaning up tap-hold", position);
+    clear_tap_hold(tap_hold);
+  }
+
+  return 0;
+}
+
+static const struct behavior_driver_api behavior_tap_hold_driver_api = {
+  .binding_pressed = on_tap_hold_binding_pressed,
+  .binding_released = on_tap_hold_binding_released,
+};
+
+int behavior_tap_hold_listener(const struct zmk_event_header *eh)
+{
+  if (!is_position_state_changed(eh)) {
+    return 0;
+  }
+  struct position_state_changed* ev = cast_position_state_changed(eh);
+  if(undecided_tap_hold == NULL) {
+    LOG_DBG("%d bubble (no undecided tap_hold active)", ev->position);
+    return 0;
+  }
+
+  if(undecided_tap_hold->position == ev->position) {
+    if(ev->state) {
+      LOG_ERR("this listener should be called before before the behavior listeners!");
+      return 0;
+    } else {
+      LOG_DBG("%d bubble undecided tap-hold keyrelease event", undecided_tap_hold->position);
+      return 0;
+    }
+  }
+
+  LOG_DBG("%d capturing %d %s event", undecided_tap_hold->position, ev->position, ev->state?"down":"up");
+
+  capture_position_event(ev);
+  if (ev->state) { 
+    decide_tap_hold(undecided_tap_hold, TH_OTHER_KEY_DOWN);
+  } else {
+    struct position_state_changed* captured_key_down = find_captured_keydown_event(ev->position); 
+    if(captured_key_down == NULL) {
+      // todo: allow key-up events for non-mod keys pressed before the TH was pressed.
+      // see scenario 3c/3d vs 3a/3b. 
```

Comment:
I wasn't sure how to implement this, as we're in a key position listener and don't know yet what kind of keypress will result here.

What is needed is something that finds the current behavior for a specific key position, so we can decide if it's a key press for a 'normal' key or something special like a momentary layer or a mod key.